### PR TITLE
Graphql client

### PR DESCRIPTION
**Pull request by idabmat**
Working WIP but still a bit rough. Todo:
- [x] add a couple more tests for custom id values
- [x] iron out exception handling
- [x] refactor duplicated code
- [x] remove unused code from the client
- [x] add documentation for new methods and classes
- [x] fix documented return values in the client
- [x] change workflow query once the fix in Alfred has landed

### CODE REVIEW

**Comment by idabmat** [comment]
@Ciboulette I'm tagging you in here for a test drive. To use this, you can use the Ruby examples and change the following in the Gemfile:
From
```
gem 'zenaton'`
```
to
```
gem 'zenaton', github: 'zenaton/zenaton-ruby', branch: 'graphql-client'
```

**Comment by Ciboulette** [comment]
> @Ciboulette I'm tagging you in here for a test drive. To use this, you can use the Ruby examples and change the following in the Gemfile:
> From
> 
> ```
> gem 'zenaton'`
> ```
> 
> to
> 
> ```
> gem 'zenaton', github: 'zenaton/zenaton-ruby', branch: 'graphql-client'
> ```

I'm testing it right now 🐘

File: lib/zenaton/services/graph_ql/create_workflow_schedule_mutation.rb [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+        end
+
+        # The variables used in the query
+        def variables
+          { 'input' => input }
+        end
+
+        private
+
+        def workflow_name
```
Comment (MrYawe) [inline lib/zenaton/services/graph_ql/create_workflow_schedule_mutation.rb]: It seems that this logic is duplicated

**Comment by Ciboulette** [comment]
@idabmat When I run `ruby launch_event.rb`

I have the following error:
```
Traceback (most recent call last):
        4: from launch_event.rb:12:in `<main>'
        3: from /usr/local/bundle/bundler/gems/zenaton-ruby-28750e3d10d2/lib/zenaton/query/builder.rb:36:in `send_event'
        2: from /usr/local/bundle/bundler/gems/zenaton-ruby-28750e3d10d2/lib/zenaton/client.rb:120:in `send_event'
        1: from /usr/local/bundle/bundler/gems/zenaton-ruby-28750e3d10d2/lib/zenaton/services/graph_ql/client.rb:79:in `send_event'
/usr/local/bundle/bundler/gems/zenaton-ruby-28750e3d10d2/lib/zenaton/services/graph_ql/client.rb:104:in `run_mutation': - ["sendEventToWorkflowByNameAndCustomId"]: Instance EventWorkflow with id: dfc19da91f is not found. (Zenaton::ExternalError)
```

File: lib/zenaton/services/graph_ql/create_workflow_schedule_mutation.rb (hunk 1 above)
Comment (idabmat) [reply lib/zenaton/services/graph_ql/create_workflow_schedule_mutation.rb]: Definitely. I'm not a fan of having this logic duplicated. Currently this method is in:
- `create_workflow_schedule_mutation.rb`
- `dispatch_workflow_mutation.rb`

I'd wait for a 3rd occurrence to have a better idea where to put this. I agree this is not the best place for it.

**Comment by idabmat** [comment]
> Good work +1
> My only concerns are these private functions in mutation classes. It seems that they are not specific to a particular query. Even if some of them are used once, it would be nice to find a better place for them. WDYT?

There are only a few mutation which have private methods at all. In all cases it seems to have to do with validation stuff. My feeling is that this is best handled at the mutation level for now as the validation logic are mutation specific for now. Ideally I'd prefer if we could move this validation to Alfred and remove it from the client libraries. WDYT ?

**Comment by idabmat** [comment]
@MrYawe After rebasing this PR on master, I noticed #87 added support for custom IDs for scheduled workflows. I've added them to this PR too in my last commit:
https://github.com/zenaton/zenaton-ruby/pull/85/commits/f0de87c09928eec0121b2d21e169865f37db4af8

I also took this opportunity to extract the duplicated logic you mentioned into a module. Probably some of that logic could live in the Workflow superclass instead, but IIRC there were some concerns about potential conflicts with user defined code.

I'm requesting a new review from you. Once this lands I'll prepare a release for the `0.6.0`.

**Pull request merged**